File: SearchInUnknownSizeArr.java
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes.
// Any problem you faced while coding this :

public class SearchInUnknownSizeArr {
    public int search(ArrayReader reader, int target) {
        int low =0;
        int high = 1;

        while(reader.get(high) < target){
            low = high;
            high = high*2;
        }

        return binarySearch (reader, low, high, target);
    }

    public int binarySearch(ArrayReader reader, int low, int high, int target){
        while(low <= high){
            int mid = low + (high - low)/2;

            if(reader.get(mid) == target){
                return mid;
            }else if(reader.get(mid) > target){
                high = mid - 1;
            }else{
                low = mid +1;
            }

        }
        return -1;
    }
}

File: search in unknown size arr.py
# """
# This is ArrayReader's API interface.
# You should not implement it, or speculate about its implementation
# """
#class ArrayReader:
#    def get(self, index: int) -> int:
#Time: O(log(n))
#Space: O(1)
class Solution:
    def search(self, reader: 'ArrayReader', target: int) -> int:
        l=0
        h=1
        lv=reader.get(l)
        hv=reader.get(h)
        while reader.get(h)<target:
            l=h
            h*=2
            
        while(l<=h):
            mid=(l+h)//2
            if reader.get(mid)<target:
                l=mid+1
            elif reader.get(mid)>target:
                h=mid-1
            else:
                return mid
        
        
        return -1
                    
